src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc

```diff
@@ -0,0 +1,241 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_placement_group_scheduler.h"
+#include <ray/protobuf/gcs.pb.h>
+#include "ray/gcs/gcs_server/gcs_placement_group_manager.h"
+
+namespace ray {
+namespace gcs {
+
+GcsPlacementGroupScheduler::GcsPlacementGroupScheduler(
+    boost::asio::io_context &io_context,
+    std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage,
+    const gcs::GcsNodeManager &gcs_node_manager,
+    std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_failure_handler,
+    std::function<void(std::shared_ptr<GcsPlacementGroup>)> schedule_success_handler,
+    ReserveResourceClientFactoryFn lease_client_factory)
+    : io_context_(io_context),
+      gcs_table_storage_(gcs_table_storage),
+      gcs_node_manager_(gcs_node_manager),
+      schedule_failure_handler_(std::move(schedule_failure_handler)),
+      schedule_success_handler_(std::move(schedule_success_handler)),
+      lease_client_factory_(std::move(lease_client_factory)) {
+  RAY_CHECK(schedule_failure_handler_ != nullptr && schedule_success_handler_ != nullptr);
+  scheduler.push_back(std::make_shared<GcsPackStrategy>());
+  scheduler.push_back(std::make_shared<GcsSpreadStrategy>());
+}
+
+std::unordered_map<BundleID, ClientID> GcsPackStrategy::Schedule(
+    std::vector<ray::BundleSpecification> &bundles, const GcsNodeManager &node_manager) {
+  std::unordered_map<BundleID, ClientID> schedule_map;
+  auto &alive_nodes = node_manager.GetAllAliveNodes();
+  for (size_t pos = 0; pos < bundles.size(); pos++) {
+    schedule_map[bundles[pos].BundleId()] =
+        ClientID::FromBinary(alive_nodes.begin()->second->node_id());
```

Comment:
Doesn't this put everything on the first node?

Reply:
Yep, this is a fake algorithm now.
I will change it after i have a resource view.